python/ray/data/datasource/file_based_datasource.py

```diff
@@ -156,7 +182,14 @@ def prepare_read(
         paths, filesystem = _resolve_paths_and_filesystem(paths, filesystem)
         paths, file_sizes = meta_provider.expand_paths(paths, filesystem)
         if partition_filter is not None:
-            paths = partition_filter(paths)
+            filtered_paths = partition_filter(paths)
+            if not filtered_paths:
+                raise ValueError(
+                    f"All provided and expanded paths {paths} have been filtered out "
+                    f"by the path filter {partition_filter}; please change either the "
+                    "provided paths or path partition filter."
+                )
```

Comment:
Good error.

Reply:
Actually, this may not be really easy to read because the {path}/{partition_filter} may clutter the message. 

**Please use the following structure:**

ERROR -> ACTION -> INPUT/OUTPUT.

**For example:**
```
ValueError: could not convert string to float: 'Hello World'
```

Reply:
@richardliaw Good call out! I was sloppy on this one.

How about:
```python
                raise ValueError(
                    "All provided and expanded paths have been filtered out by "
                    "the path filter; please change either the provided paths or "
                    f"the path filter.\nPaths: {paths}\nFilter: {partition_filter}"
                )
```

Reply:
Nice! better.